job-runner/chain/traverser.go

```diff
@@ -503,3 +547,68 @@ PROCESS_TO_VISIT_LIST:
 
 	return completedJobs
 }
+
+// Stops the traverser and sends a Suspended Job Chain (SJC) to the Request Manager.
+// The SJC contains all the necessary info to recreate the traverser and resume
+// running the job chain.
+// suspend is called when the Job Runner is shutting down.
+func (t *traverser) suspend() {
+	t.logger.Info("suspending job chain")
+	defer t.logger.Info("job chain suspended")
+
+	// Stop all active runners.
+	activeRunners, err := t.runnerRepo.Items()
+	if err != nil {
+		// Log error but continue trying to suspend the job chain - JR is being
+		// shut down. (even on error, activeRunners won't be set to nil)
+		t.logger.Errorf("problem retrieving job runners from repo: %s", err)
+	}
+	for jobId, runner := range activeRunners {
+		err := runner.Stop() // this should return quickly
+		if err != nil {
+			t.logger.Errorf("problem stopping job runner (job id = %s): %s", jobId, err) // log error but continue
+		}
+	}
+
+	// Wait for chain to finish running - timeout after 10 seconds.
+	waitChan := time.After(10 * time.Second)
+	select {
+	case <-waitChan:
+	case <-t.doneChan:
+	}
+
+	if t.chain.State() != proto.STATE_COMPLETE {
+		// Only send SJC if chain didn't complete successfully.
+		// If chain completed successfully, Run() will have told the RM.
+
+		// Save chain state (in case there's an issue sending the SJC to the RM).
+		t.chain.SetState(proto.STATE_SUSPENDED)
+		t.chainRepo.Set(t.chain)
+
+		// Create a SuspendedJobChain and send it to the RM.
+		jobChain := t.chain.JobChain()
+		requestId := jobChain.RequestId
+		sequenceRetries := t.chain.SequenceRetryCounts()
+		sjc := proto.SuspendedJobChain{
+			RequestId:       requestId,
+			JobChain:        jobChain,
+			JobTries:        t.jobTries,
+			StoppedJobTries: t.stoppedJobTries,
+			SequenceRetries: sequenceRetries,
+		}
+		if err := t.rmc.SuspendRequest(requestId, sjc); err != nil {
+			t.logger.Errorf("problem sending suspended job chain to RM: %s", err)
+		} else {
+			t.chainRepo.Remove(t.chain.RequestId())
+		}
+	}
+}
+
+func (t *traverser) suspended() bool {
```

Comment:
For consistency, I'd rename to `shuttingDown` like the func in api since it's the same code/test on `shutdownChan`. Plus, the JR Itself isn't really suspended, only job chains are suspended.

Reply:
Hm, one second reading maybe this is ok. Still thinking about it. It helps to think of traverser + chain as one thing.